src/rcc/clocking.rs

```diff
@@ -0,0 +1,292 @@
+            rcc.bdcr
+                .modify(|_, w| unsafe { w.lsedrv().bits(0b11).lseon().set_bit() });
+            while rcc.bdcr.read().lserdy().bit_is_clear() {}
+            rcc.cr.modify(|_, w| w.msipllen().set_bit());
+        }
+        (self.freq(), 0b00)
+    }
+}
+
+impl InputClock for MediumSpeedInternalRC {
+    fn freq(&self) -> u32 {
+        self.freq
+    }
+}
+
+/// High-speed external 4-48 MHz oscillator
+#[derive(Clone, Copy)]
+pub struct HighSpeedExternalOSC(pub u32);
+impl InputClock for HighSpeedExternalOSC {
+    fn freq(&self) -> u32 {
+        self.0
+    }
+}
+
+impl HighSpeedExternalOSC {
+    /// Turns on the HSE oscillator.
+    ///
+    /// (Should this also configure the pin?)
+    pub fn configure(&self, rcc: &rcc::RegisterBlock) -> (u32, u8) {
+        rcc.cr.modify(|_, w| w.hseon().set_bit());
+        while rcc.cr.read().hserdy().bit_is_clear() {}
+        (self.freq(), 0b10)
+    }
+}
+
+pub enum RtcClkSource {
+    None,
+    /// Internal 32 kHz RC
+    LSI,
+    /// External 32.768 kHz oscillator
+    LSE,
+    /// High-speed external oscillator, prescaled by (a fixed value of) 32
+    HSEDiv32,
+}
+
+    impl RtcClkSource {
+        /// Returns the output frequency of the RtcClkSource based on its input.
+        pub fn freq(&self, hse: Option<HighSpeedExternalOSC>) -> Option<u32> {
+            match *self {
+                RtcClkSource::None => None,
+                RtcClkSource::LSI => Some(32_000),
+                RtcClkSource::LSE => Some(32_768),
+                RtcClkSource::HSEDiv32 => {
+                    if let Some(clk) = hse {
+                        Some(clk.freq() / 32)
+                    } else {
+                        None
+                    }
+                }
+            }
+        }
+
+        pub fn bits(&self) -> u8 {
+            match *self {
+                RtcClkSource::None => 0,
+                RtcClkSource::LSI => 1,
+                RtcClkSource::LSE => 2,
+                RtcClkSource::HSEDiv32 => 3,
+            }
+        }
+    }
+
+#[derive(Clone, Copy)]
+pub enum SysClkSource {
+    HSI16(HighSpeedInternal16RC),
+    MSI(MediumSpeedInternalRC),
+    HSE(HighSpeedExternalOSC),
+    PLL(PLLClkOutput),
+}
+
+impl InputClock for SysClkSource {
+    fn freq(&self) -> u32 {
+        match *self {
+            SysClkSource::HSI16(_) => 16_000_000,
+            SysClkSource::MSI(s) => s.freq(),
+            SysClkSource::HSE(s) => s.freq(),
+            SysClkSource::PLL(s) => s.freq(),
+        }
+    }
+}
+
+/// PLLCLK output of PLL module
+#[derive(Clone, Copy)]
+pub struct PLLClkOutput {
+    pub src: PLLClkSource,
+    pub m: u8,
+    n: u8,
+    r: u8,
+    f: u32,
+}
+
+impl PLLClkOutput {
+    /// Create a new PLL clock source to use as an input. The arguments refer to the scale
+    /// factors described in Figs. 15 and 16 of the reference manual, and end up in the PLLM,
+    /// PLLN, and PLLR fields of the PLLCFGR register.
```

Comment:
Please add note that function will panic if resulting frequency exceeds SYS_CLOCK_MAX